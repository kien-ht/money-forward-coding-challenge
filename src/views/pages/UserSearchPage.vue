<template>
  <MainLayout>
    <b-input-group size="md" class="account-input-group mb-3">
      <b-form-input
        v-model="textSearch"
        v-focus
        :state="rule"
        placeholder="Type user_id to search"
        aria-describedby="input-live-help input-live-feedback"
        @keyup.enter="onClickSearchButton"
      ></b-form-input>
      <b-input-group-append>
        <b-button
          size="md"
          text="Button"
          variant="dark"
          :disabled="!rule"
          @click="onClickSearchButton"
          >Search</b-button
        >
      </b-input-group-append>
      <b-form-invalid-feedback id="input-live-feedback">
        You can't search with an empty text!
      </b-form-invalid-feedback>
    </b-input-group>
  </MainLayout>
</template>

<script land="ts">
import Vue from "vue";
import MainLayout from "@/views/layouts/MainLayout.vue";

export default Vue.extend({
  name: "UserSearchPage",

  components: {
    MainLayout,
  },

  directives: {
    focus: {
      inserted(el, binding, vnode) {
        el.focus();
      },
    },
  },

  data() {
    return {
      textSearch: "",
      isTouched: false,
    };
  },

  computed: {
    rule() {
      if (!this.isTouched) return null;
      return Boolean(this.textSearch);
    },
  },

  created() {
    this.setupSearchInputWatcher();
  },

  methods: {
    // region [EVENT]
    async onClickSearchButton() {
      await this.$store.dispatch(
        "setCurrentSearchUserAccountList",
        this.textSearch
      );

      if (!this.$store.getters.currentSearchUserAccountList.length) {
        return this.$root.notify.open({
          title: "Notice",
          body: "No user is found!",
        });
      }
      this.$router.push(`/users/${this.textSearch}`);
    },
    // end region

    // region [PRIVATE]
    setupSearchInputWatcher() {
      const unwatch = this.$watch("textSearch", () => {
        this.isTouched = true;
        unwatch();
      });
    },
    // end region
  },
});
</script>

<style lang="scss" scoped>
.account-input-group {
  text-align: left;
}
</style>
