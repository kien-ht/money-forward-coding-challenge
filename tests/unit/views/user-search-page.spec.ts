import { shallowMount, mount } from "@vue/test-utils";
import UserSearchPage from "@/views/pages/UserSearchPage.vue";

describe("UserSearchPage.vue", () => {
  it("snapshot", () => {
    const wrapper = mount(UserSearchPage, {
      stubs: ['router-link', 'router-view']
    })
    expect(wrapper).toMatchSnapshot()
  })

  let wrapper: any

  beforeEach(() => {
    wrapper = shallowMount(UserSearchPage, {})
  })

  it("renders search input", () => {
    const input = wrapper.findComponent({ name: 'b-form-input' })
    expect(input.exists()).toBe(true)
  });

  it("renders search input placeholder", () => {
    expect(wrapper.findComponent({ name: 'b-form-input' }).props().placeholder).toBe('Type user_id to search')
  });

  it("renders search button", () => {
    expect(wrapper.findComponent({ name: 'b-button' }).text()).toBe("Search")
  });

});
