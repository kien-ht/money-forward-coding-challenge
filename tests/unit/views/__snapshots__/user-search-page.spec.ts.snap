// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSearchPage.vue snapshot 1`] = `
<div class="layout-wrapper">
  <div class="page-topbar bg-white">
    <router-link-stub to="/"><span><img src="@/assets/money-forward-logo.png" alt="" height="50"></span></router-link-stub>
  </div>
  <div class="h-100 container-fluid">
    <div class="row h-100">
      <div class="bg-warning py-5 px-4 d-none d-md-block col-3">
        <ul class="nav nav-pills flex-column"><button type="button" class="btn w-100 btn-dark">Account Search</button></ul>
      </div>
      <div class="p-5 col">
        <div role="group" class="input-group account-input-group mb-3 input-group-md">
          <!----><input type="text" placeholder="Type user_id to search" class="form-control" aria-describedby="input-live-help input-live-feedback">
          <div class="input-group-append"><button text="Button" type="button" disabled="disabled" class="btn btn-dark btn-md disabled">Search</button></div>
          <div id="input-live-feedback" class="invalid-feedback">
            You can't search with an empty text!
          </div>
          <!---->
        </div>
      </div>
    </div>
  </div>
</div>
`;
